llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -2553,6 +2553,48 @@ void CombinerHelper::applyCombineTruncOfExt(
   MI.eraseFromParent();
 }
 
+bool CombinerHelper::matchCombineShlOfAnd(MachineInstr &MI, Register &Reg) {
+  // We're trying to match the following pattern:
+  //   %t1 = G_AND %x, (2^imm - 1)
+  //   %root = G_SHL %t1, imm
+  // -->
+  //   %root = G_SHL %x, imm
+  // Where imm >= half size of %x
+  assert(MI.getOpcode() == TargetOpcode::G_SHL && "Expected a G_SHL");
+
+  const Register SrcReg = MI.getOperand(1).getReg();
+  const Register ShftReg = MI.getOperand(2).getReg();
+  const LLT SrcTy = MRI.getType(SrcReg);
+  const unsigned Size = SrcTy.getSizeInBits();
+
+  const auto Shift = getIConstantVRegSExtVal(ShftReg, MRI);
+  if (!Shift)
+    return false;
+
+  const unsigned ShiftVal = Shift.value();
+  if (ShiftVal >= Size / 2) {
```

Comment:
nit: sometimes is better to invert the `if` to have an early return here (reduce indentation).